File: apps/api/app/schemas/org.py
from __future__ import annotations

import datetime
import uuid
from enum import Enum

from pydantic import BaseModel, ConfigDict, EmailStr, Field


class Role(str, Enum):
    owner = "owner"
    admin = "admin"
    editor = "editor"
    viewer = "viewer"


# Roles that can be assigned through invitations / role changes. ``owner`` is
# never assignable via the API (it is granted only to the org creator).
ASSIGNABLE_ROLES = {Role.admin, Role.editor, Role.viewer}
# Roles with administration rights.
ADMIN_ROLES = {Role.owner.value, Role.admin.value}


class InvitableRole(str, Enum):
    admin = "admin"
    editor = "editor"
    viewer = "viewer"


# --- Organization -----------------------------------------------------------


class OrgOut(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    id: uuid.UUID
    name: str
    siren: str | None = None
    sector: str | None = None
    country: str | None = None
    company_size: str | None = None
    dpo_name: str | None = None
    settings: dict[str, object]


class OrgUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=255)
    siren: str | None = Field(default=None, max_length=32)
    sector: str | None = Field(default=None, max_length=120)
    country: str | None = Field(default=None, max_length=120)
    company_size: str | None = Field(default=None, max_length=64)
    dpo_name: str | None = Field(default=None, max_length=255)


class OrgSettingsUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    llm_provider: str | None = None
    di_threshold: float | None = Field(default=None, ge=0.0, le=1.0)
    retention_days: int | None = Field(default=None, ge=1, le=3650)
    report_options: dict[str, object] | None = None


# --- Profile (me) -----------------------------------------------------------


class MeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    id: uuid.UUID
    email: str
    first_name: str | None = None
    role: str
    org_id: uuid.UUID


class MeUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    first_name: str | None = Field(default=None, max_length=120)


# --- Members ----------------------------------------------------------------


class MemberOut(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    id: uuid.UUID
    email: str
    first_name: str | None = None
    role: str
    created_at: datetime.datetime


class MemberRoleUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    role: Role


# --- Invitations ------------------------------------------------------------


class InvitationCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    email: EmailStr
    role: InvitableRole


class InvitationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    id: uuid.UUID
    email: str
    role: str
    status: str
    created_at: datetime.datetime
    expires_at: datetime.datetime


class InvitationCreateOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    invitation: InvitationOut
    invite_url: str
    email_sent: bool

File: apps/web/lib/api/org.ts
import { api } from '@/lib/api/client';
import type { Sector } from '@/lib/api/audits';

export type { Sector };

/* ─── Rôles & types ──────────────────────────────────────────────────────── */

/** Rôle d'un membre dans l'organisation (RBAC backend). */
export type Role = 'owner' | 'admin' | 'editor' | 'viewer';

/** Rôles assignables lors d'une invitation (l'owner ne s'invite pas). */
export type InvitableRole = 'admin' | 'editor' | 'viewer';

/* ─── Organisation ───────────────────────────────────────────────────────── */

/** Options de rapport persistées dans `org.settings.report_options`. */
export type ReportOptions = Record<string, unknown>;

export type OrgSettings = {
  llm_provider?: string | null;
  di_threshold?: number | null;
  retention_days?: number | null;
  report_options?: ReportOptions | null;
  [key: string]: unknown;
};

export type OrgOut = {
  id: string;
  name: string;
  siren?: string | null;
  sector?: Sector | null;
  country?: string | null;
  company_size?: string | null;
  dpo_name?: string | null;
  settings: OrgSettings;
};

export type OrgUpdate = {
  name?: string;
  siren?: string;
  sector?: Sector;
  country?: string;
  company_size?: string;
  dpo_name?: string;
};

export type OrgSettingsUpdate = {
  llm_provider?: string;
  /** 0..1 */
  di_threshold?: number;
  /** 1..3650 */
  retention_days?: number;
  report_options?: ReportOptions;
};

export async function fetchOrg(): Promise<OrgOut> {
  const { data } = await api.get<OrgOut>('/org');
  return data;
}

export async function updateOrg(body: OrgUpdate): Promise<OrgOut> {
  const { data } = await api.patch<OrgOut>('/org', body);
  return data;
}

export async function updateOrgSettings(
  body: OrgSettingsUpdate,
): Promise<OrgOut> {
  const { data } = await api.patch<OrgOut>('/org/settings', body);
  return data;
}

/* ─── Compte courant ─────────────────────────────────────────────────────── */

export type MeOut = {
  id: string;
  email: string;
  first_name?: string | null;
  role: Role;
  org_id: string;
};

export type MeUpdate = {
  first_name?: string;
};

export async function fetchMe(): Promise<MeOut> {
  const { data } = await api.get<MeOut>('/me');
  return data;
}

export async function updateMe(body: MeUpdate): Promise<MeOut> {
  const { data } = await api.patch<MeOut>('/me', body);
  return data;
}

/* ─── Membres ────────────────────────────────────────────────────────────── */

export type MemberOut = {
  id: string;
  email: string;
  first_name?: string | null;
  role: Role;
  created_at: string;
};

export async function fetchMembers(): Promise<MemberOut[]> {
  const { data } = await api.get<MemberOut[]>('/org/members');
  return data;
}

export async function updateMemberRole(
  userId: string,
  role: Role,
): Promise<MemberOut> {
  const { data } = await api.patch<MemberOut>(`/org/members/${userId}`, {
    role,
  });
  return data;
}

/* ─── Invitations ────────────────────────────────────────────────────────── */

export type InvitationStatus = 'pending' | 'accepted' | 'expired' | 'revoked';

export type InvitationOut = {
  id: string;
  email: string;
  role: InvitableRole;
  status: InvitationStatus;
  created_at: string;
  expires_at: string;
};

export type InvitationCreate = {
  email: string;
  role: InvitableRole;
};

export type InvitationCreateResult = {
  invitation: InvitationOut;
  invite_url: string;
  email_sent: boolean;
};

export async function fetchInvitations(): Promise<InvitationOut[]> {
  const { data } = await api.get<InvitationOut[]>('/org/invitations');
  return data;
}

export async function createInvitation(
  body: InvitationCreate,
): Promise<InvitationCreateResult> {
  const { data } = await api.post<InvitationCreateResult>(
    '/org/invitations',
    body,
  );
  return data;
}

export async function revokeInvitation(id: string): Promise<void> {
  await api.delete(`/org/invitations/${id}`);
}

/* ─── Helpers d'affichage / erreurs ──────────────────────────────────────── */

/** Libellé FR du rôle pour l'UI (owner et admin partagent « Administrateur »). */
export const ROLE_LABELS: Record<Role, string> = {
  owner: 'Administrateur',
  admin: 'Administrateur',
  editor: 'Éditeur',
  viewer: 'Lecture seule',
};

/** Un utilisateur owner/admin peut administrer l'organisation. */
export function canManageOrg(role: Role | undefined): boolean {
  return role === 'owner' || role === 'admin';
}

/**
 * Extrait le message d'erreur lisible d'une réponse `application/problem+json`
 * (champ `detail`), avec repli sur le message d'exception.
 */
export function problemDetail(err: unknown, fallback: string): string {
  if (err && typeof err === 'object') {
    const resp = (err as { response?: { data?: { detail?: unknown } } }).response;
    const detail = resp?.data?.detail;
    if (typeof detail === 'string' && detail.trim()) return detail;
  }
  return fallback;
}
